File: leetcode/283-moving-zeros/main.go
package main

/*
	1st approach:
	- 2 index pointers: slow pointer points to the first zero, fast pointer is for iteration
	- swap the non-zero number with the zero

	Time		O(n)
	Space		O(1)
	64 ms, faster than 89.29%
*/
func moveZeroes(nums []int) {
	idx0 := -1
	for i := 0; i < len(nums); i++ {
		if nums[i] != 0 {
			nums[idx0+1], nums[i] = nums[i], nums[idx0+1]
			idx0++
		}
	}
}

func main() {

}

File: leetcode/283-moving-zeros/main.py
"""
    questions to ask:
    - in-place? yes
    - will there be an empty array? yes
"""


"""
    1st approach: 2 pointers
    - wap the values until we move all the zero to the back of the array
"""


class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        # always points to the first zero
        slow = 0
        for i in range(len(nums)):
            # if the current number is not zero, swap with the value at slow
            if nums[i] != 0:
                nums[i], nums[slow] = nums[slow], nums[i]
                slow += 1
